secrets/secrets.go

```diff
@@ -0,0 +1,95 @@
+// Copyright 2023 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package secrets
+
+import (
+	"context"
+
+	"github.com/go-kit/log"
+)
+
+// Secret represents a sensitive value.
+type Secret interface {
+	// Fetch fetches the secret content.
+	Fetch(ctx context.Context) (string, error)
+}
+
+// SecretFn wraps a function to make it a Secret.
+type SecretFn func(ctx context.Context) (string, error)
+
+func (fn *SecretFn) Fetch(ctx context.Context) (string, error) {
+	return (*fn)(ctx)
+}
+
+// Provider is a secret provider.
+type Provider[T any] interface {
```

Comment:
not a blocker for fork: Something is odd here. I don't know if generics here are really necessary. We could equally pass un-serialized config values and return interfaces (e.g. Provider[T] in Config[T])